packages/doenetml-worker-rust/lib-doenetml-core/src/core/state/types/math_expr.rs

```diff
@@ -0,0 +1,429 @@
+    ///     "2 x y + 2 x + y + 1"
+    /// );
+    /// ```
+    pub fn normalize(&self, params: NormalizeParams) -> MathExpr {
+        let math_object = match normalize_math(&self.math_object, params) {
+            Ok(res) => res,
+            Err(_) => BLANK_MATH_OBJECT.to_owned(),
+        };
+
+        MathExpr { math_object }
+    }
+
+    /// Create a new mathematical expression by adding `term` to the current expression.
+    pub fn add(&self, term: MathOrPrimitive) -> MathExpr {
+        let term = serde_json::to_string(&term).unwrap();
+
+        let js_source = format!(
+            r#"JSON.parse('{}', serializedComponentsReviver).add(JSON.parse('{}', serializedComponentsReviver));"#,
+            self.math_object, term
+        );
+
+        let math_object = match eval_js(&js_source) {
+            Ok(res) => res,
+            Err(_) => BLANK_MATH_OBJECT.to_owned(),
+        };
+
+        MathExpr { math_object }
+    }
+
+    /// Create a new mathematical expression by subtracting `term` to the current expression.
+    pub fn subtract(&self, term: MathOrPrimitive) -> MathExpr {
+        let term = serde_json::to_string(&term).unwrap();
+
+        let js_source = format!(
+            r#"JSON.parse('{}', serializedComponentsReviver).subtract(JSON.parse('{}', serializedComponentsReviver));"#,
+            self.math_object, term
+        );
+
+        let math_object = match eval_js(&js_source) {
+            Ok(res) => res,
+            Err(_) => BLANK_MATH_OBJECT.to_owned(),
+        };
+
+        MathExpr { math_object }
+    }
+
+    /// Create a new mathematical expression by multiplying the current expression by `term`.
+    pub fn multiply(&self, factor: MathOrPrimitive) -> MathExpr {
+        let factor = serde_json::to_string(&factor).unwrap();
+
+        let js_source = format!(
+            r#"JSON.parse('{}', serializedComponentsReviver).multiply(JSON.parse('{}', serializedComponentsReviver));"#,
+            self.math_object, factor
+        );
+
+        let math_object = match eval_js(&js_source) {
+            Ok(res) => res,
+            Err(_) => BLANK_MATH_OBJECT.to_owned(),
+        };
+
+        MathExpr { math_object }
+    }
+
+    /// Create a new mathematical expression by dividing the current expression by `term`.
+    pub fn divide(&self, factor: MathOrPrimitive) -> MathExpr {
+        let factor = serde_json::to_string(&factor).unwrap();
+
+        let js_source = format!(
+            r#"JSON.parse('{}', serializedComponentsReviver).divide(JSON.parse('{}', serializedComponentsReviver));"#,
+            self.math_object, factor
+        );
+
+        let math_object = match eval_js(&js_source) {
+            Ok(res) => res,
+            Err(_) => BLANK_MATH_OBJECT.to_owned(),
+        };
+
+        MathExpr { math_object }
+    }
+}
+
+/// An enum of different types of values that can be interpreted as mathematical expressions.
+#[derive(Debug)]
+pub enum MathOrPrimitive {
```

Comment:
I don't like names that spell out the type, in general. Ideally the name should reflect the objects use and properties, since we can always figure out the type via a mouse hover.

What about `Math`? You already have `MathExpr` for something else, otherwise that might be a good name...

Reply:
`MathOrPrimitive` doesn't work by itself. It is designed for arguments of `MathExpr` functions like `.add()` to give a shortcut so that you can, for example, add `1` without first creating a `MathExpr` for `1`.

OK, just writing out this reply helped me think through it. I'm thinking I'll call it `MathArg` and document it with the usage I just described.